### DISCUSSION TITLE
Flattening nullable nested objects

### DISCUSSION THREAD

**Discussion by jobadder-daviddancy** [riok/mapperly#355] [Q&A]
Hi all,

I've been encouraged to look at Mapperly to reduce the amount of custom mapping code I'm writing for a project.

The code it generates currently does not appear to null-check any lower than the first level of a nested object that's being flattened, which causes a `NullReferenceException` when a null is encountered.

For example from the docs, 

`[MapProperty(Source = new[] { nameof(Car), nameof(Car.Make), nameof(Car.Make.Id) }, Target = new[] { nameof(CarDto), nameof(CarDto.MakeId) })]`

If `car.Make.Id` is null this will throw.

Is there any workaround for this problem? Ideally I'd like the code-gen to implement something like

`car?.Make?.Id ?? default` when `Make` and `Id` are themselves nullable.

Thanks!

**Comment by jobadder-daviddancy** [comment]
Please ignore - sorry for the noise!

I double-checked my code and it turns out some of the source properties in question were nullable but assigned to `default!` on initialisation. I've removed that and the code-gen looks like it now takes account of the nullability.

**Comment by Kog008** [comment]
What do I miss during flattening? In my opinion I have the same structure, like in the docs. Two EF Core 6 entity classes.
The parent class:
```
public partial class Product
{
    public byte PlantId { get; set; }
    public byte ProductId { get; set; }
    public string ProductName { get; set; } = null!;
    public float NominaleDicke { get; set; }

    public virtual Plant Pl { get; set; } = null!;
}
```
The child class:
```
public partial class Plant
{
  public byte PlantId { get; set; }
  public string PlantName { get; set; } = null!;
  public string PlantDescription { get; set; } = null!;
}
```
I want to flatten it to a ProductDto class:
```
public partial class ProductDto
{
  // properties from Product class
  public byte PlantId { get; set; }
  public byte ProductId { get; set; }
  public string ProductName { get; set; } = null!;
  public float NominaleDicke { get; set; }
  
  // properties from Plant class
  public string PlantName { get; set; } = null!;
  public string PlantDescription { get; set; } = null!;
}
```
If I use pascal case notation in the dto class it works like intended, but it looks pretty stupid, having properties `PlPlantName`, or `PlantPlantName`. So I tried the MapProperty decorator:
```
[MapProperty(
	source: new[] { nameof(Product), nameof(Product.Pl), nameof(Product.Pl.PlantName) },
	target: new[] { nameof(ProductDto), nameof(ProductDto.PlantName) }
)]
[MapProperty(
	source: new[] { nameof(Product), nameof(Product.Pl), nameof(Product.Pl.PlantDescription) },
	target: new[] { nameof(ProductDto), nameof(ProductDto.PlantDescription) }
)]
public partial ProductDto ToProductDto(Product p);
```

But that leads into errors: 

- error RMG005: Specified property ProductDto.PlantName on mapping target type EgsPlattformBackend.Dtos.F85.trv.ProductDto was not found
- error RMG005: Specified property ProductDto.PlantDescription on mapping target type EgsPlattformBackend.Dtos.F85.trv.ProductDto was not found

I tried several, different approaches, but ran into the same errors, or the mapping was just null. Does somebody see my mistake? What the heck do I oversee? ^^

**Reply by latonz** [reply]
You only need to specify the full path of the property, but it seems that you include the class. Eg. replace
```c#
[MapProperty(
	source: new[] { nameof(Product), nameof(Product.Pl), nameof(Product.Pl.PlantName) },
	target: new[] { nameof(ProductDto), nameof(ProductDto.PlantName) }
)]
```
with
```c#
[MapProperty(
	source: new[] { nameof(Product.Pl), nameof(Product.Pl.PlantName) },
	target: new[] { nameof(ProductDto.PlantName) }
)]
```
